src/target/rp.c

```diff
@@ -554,56 +611,216 @@ static void rp_spi_read(
 }
 
-static uint32_t rp_get_flash_length(target *t)
+// Connect the XIP controller to the flash pads
+static void rp_flash_connect_internal(target *t)
 {
-	uint32_t size = MAX_FLASH;
-	uint32_t bootsec[16];
-	size_t i;
-
-	rp_flash_resume(t);
-	target_mem_read(t, bootsec, RP_XIP_FLASH_BASE, sizeof(bootsec));
-	for (i = 0; i < 16; i++) {
-		if ((bootsec[i] != 0x00) && (bootsec[i] != 0xff))
-			break;
-	}
+    // Use hard reset to force IO and pad controls to known state (don't touch
+    // IO_BANK0 as that does not affect XIP signals)
+	uint32_t reset = target_mem_read32(t, RP_RESETS_RESET);
+    target_mem_write32(t, RP_RESETS_RESET, reset | RP_RESETS_RESET_IO_QSPI_BITS | RP_RESETS_RESET_PADS_QSPI_BITS);
+    target_mem_write32(t, RP_RESETS_RESET, reset);
+	while (~target_mem_read32(t, RP_RESETS_RESET_DONE) & (RP_RESETS_RESET_IO_QSPI_BITS | RP_RESETS_RESET_PADS_QSPI_BITS));
+
+    // Then mux XIP block onto internal QSPI flash pads
+	target_mem_write32(t, RP_GPIO_QSPI_SCLK_CTRL, 0);
+	target_mem_write32(t, RP_GPIO_QSPI_CS_CTRL, 0);
+	target_mem_write32(t, RP_GPIO_QSPI_SD0_CTRL, 0);
+	target_mem_write32(t, RP_GPIO_QSPI_SD1_CTRL, 0);
+	target_mem_write32(t, RP_GPIO_QSPI_SD2_CTRL, 0);
+	target_mem_write32(t, RP_GPIO_QSPI_SD3_CTRL, 0);
+}
 
-	if (i < 16) {
-		// We have some data (hopefully a valid program) stored in the start
-		// of the flash memory. We can check if the start of this data is
-		// mirrored anywhere else in the flash as the flash region will repeat
-		// when we try to read out of bounds.
-		uint32_t mirrorsec[16];
-		while (size > FLASHSIZE_4K_SECTOR) {
-			target_mem_read(t, mirrorsec, RP_XIP_FLASH_BASE + size, sizeof(bootsec));
-			if (memcmp(bootsec, mirrorsec, sizeof(bootsec)) != 0)
-				return size << 1U;
-			size >>= 1U;
-		}
-	}
+// Set up the SSI controller for standard SPI mode,i.e. for every byte sent we get one back
+// This is only called by flash_exit_xip(), not by any of the other functions.
+// This makes it possible for the debugger or user code to edit SPI settings
+// e.g. baud rate, CPOL/CPHA.
+static void rp_flash_init_spi(target *t)
+{
+	// Disable SSI for further config
+	target_mem_write32(t, RP_SSI_ENABLE, 0);
+	// Clear sticky errors (clear-on-read)
+	target_mem_read32(t, RP_SSI_SR);
+	target_mem_read32(t, RP_SSI_ICR);
+	// Hopefully-conservative baud rate for boot and programming
+	target_mem_write32(t, RP_SSI_BAUD, 6);
+	target_mem_write32(t, RP_SSI_CTRL0,
+			RP_SSI_CTRL0_FRF_SERIAL |	// Standard 1-bit SPI serial frames
+			RP_SSI_CTRL0_DATA_BITS(8) |	// 8 clocks per data frame
+			RP_SSI_CTRL0_TMOD_BIDI		// TX and RX FIFOs are both used for every byte
+	);
+	// Slave selected when transfers in progress
+	target_mem_write32(t, RP_SSI_SER, 1);
+	// Re-enable
+	target_mem_write32(t, RP_SSI_ENABLE, 1);
+}
+
+// Also allow any unbounded loops to check whether the above abort condition
+// was asserted, and terminate early
+static int rp_flash_was_aborted(target *t) {
```

Comment:
The { here belongs on a new line per the codebase style

Reply:
Fixed!